core/src/main/java/org/apache/druid/timeline/partition/HashBasedNumberedShardSpec.java

```diff
@@ -90,10 +98,39 @@ public int getNumBuckets()
     return partitionDimensions;
   }
 
+  @Override
+  public List<String> getDomainDimensions()
+  {
+    return partitionDimensions;
+  }
+
   @Override
   public boolean isInChunk(long timestamp, InputRow inputRow)
   {
-    return (((long) hash(timestamp, inputRow)) - bucketId) % numBuckets == 0;
+    return Math.abs(hash(timestamp, inputRow) % numBuckets) == bucketId % numBuckets;
+  }
+
+  /**
+   * Check if the current segment possibly holds records if the values of dimensions in {@link #partitionDimensions}
+   * are of {@code partitionDimensionsValues}
+   *
+   * @param partitionDimensionsValues An instance of values of dimensions in {@link #partitionDimensions}
+   *
+   * @return Whether the current segment possibly holds records for the given values of partition dimensions
+   */
+  private boolean isInChunk(Map<String, String> partitionDimensionsValues)
+  {
+    assert !partitionDimensions.isEmpty();
+    List<Object> groupKey = Lists.transform(
+        partitionDimensions,
+        o -> Collections.singletonList(partitionDimensionsValues.get(o))
+    );
+    try {
+      return Math.abs(hash(jsonMapper, groupKey) % numBuckets) == bucketId % numBuckets;
```

Comment:
Can we extract `Math.abs(hash % numBuckets) == bucketId % numBuckets` as a common method so that we can avoid a [mistake that you have fixed](https://github.com/apache/druid/pull/9810#discussion_r418961446) in the future?

Reply:
Good call! Changed. Actually I have to extract only `Math.abs(hash % numBuckets)` because in the function used by ingestion `static ShardSpecLookup createHashLookup` only needs `Math.abs(hash % numBuckets)` instead of the whole statement you mentioned.

Reply:
:+1: